### ISSUE TITLE
BarChart: non equidistand data and last data point not drawn

### ISSUE THREAD

**Issue by osaidmakhalfeh** [GSI-CS-CO/chart-fx#265]
Hi Dears,

I'm facing a problem while drawing a barChart with small values for the yAxis.

here is a snapshot for the values I'm trying to draw:

![asas](https://user-images.githubusercontent.com/44381968/94153323-4fbb4580-fe85-11ea-98de-d6d6b5a3b65b.png)

Unfortunately, the result looks like that there is one bar rendered, but it took the full width of the screen, here is a snapshot of it.

![floatissue](https://user-images.githubusercontent.com/44381968/94153931-09b2b180-fe86-11ea-9db9-4e799b93d337.PNG)

here is my sample to share the knowledge.
https://github.com/osaidmakhalfeh/testChartsProject/blob/master/pcCharts/src/main/java/BarChart.java

Thanks again.

**Comment by osaidmakhalfeh** [comment]
@RalphSteinhagen Sorry for bothering you my friend, but please I have a question, I am using the same solution as the previous problem. I had a problem when using small numbers. the bars not rendered as expected. Can you help me?
Thank you

**Comment by RalphSteinhagen** [comment]
@osaidmakhalfeh you seem to have non-equidistant data in your bar-chart (notably the point around x=20 which is about 15 units away from its neighbour).

The relative-width-based bar renderer cannot easily handle this (width ~ 1/2 distance to neighbouring point). You may want to use either a fixed spacing between x-values or a custom bar width:

```Java
   errorRenderer2.setDrawBars(true);
    errorRenderer2.setDynamicBarWidth(false);
    errorRenderer2.setBarWidth(5);
```

yielding something like:
![image](https://user-images.githubusercontent.com/46007894/94185985-602fe800-fea6-11ea-83b0-5383369f4456.png)

Alternatively, you may want to check out the histogram samples and the ErrorDataSetRenderingStylingSample in particular. The latter has a UI where you can play with the various settings...

May I ask what your targeted application/use-case is? Sometimes this helps to get a feeling for what you would like to achieve. Thanks in advance and hope the above helps.

**Comment by osaidmakhalfeh** [comment]
@RalphSteinhagen Thank you for your response again my amazing friend.

x -> [12.6, 12.5, 12.4, 12.3, 12.2]      
y -> [1.001650853992342, 1.0000024010700712, 0.9977828294515343, 0.9974950907069192]

I tried to draw the above data on the same example project. Unfortunately, there is one problem while drawing the minimum and the maximum points. I do not know why are these two points, not rendered. 
to solve this, I did something not professional, I added two dummy data points on the two sides to handle my target points.

For the app I'm working on, it's an app that reads data from a hardware sensor and handles some equations. We may use this data to do some nature-specific statistics.

**Comment by RalphSteinhagen** [comment]
@osaidmakhalfeh the default ErrorDataSetRenderer setting performs a data reduction in case data points overlap i.e. if they are drawn on or nearly on the same pixel (the minimum distance is configurable).

If you have only a few data points (ie. < 1k) and do not want this point reduction feature, you may disable this by `ErrorDataSetRenderer::setPointReduction(boolean)`. This might be useful in your case because your sample seems to have in addition non-equidistant data point distances.

**Comment by wirew0rm** [comment]
I was able to reproduce the issue with the ErrorDataSetRendererStyling sample, the inverted x-Axis seems to prevent the first and last point from being drawn.
I also noticed, that the bars vanish when panning before they would actually leave the viewport, so I guess somehow the computation of the first and last point to display gets confused when the axis is inverted.
I tried to set `renderer.setAssumeSorted(false)`which only lead to all points vanishing.
It also is not limited to the bar chart, but applies to all types of charts with inverted x axis. I located the bug in the ErrorDataSetRenderer and will provide a PR with a fix shortly.
